lib/widgets/action_sheet.dart

```diff
@@ -183,41 +147,158 @@ class MessageActionSheetCancelButton extends StatelessWidget {
+
+  final Message message;
+}
+
+class ReactionButtons extends StatelessWidget {
+  const ReactionButtons({
+    super.key,
+    required this.message,
+    required this.pageContext,
+  });
+
+  final Message message;
+
+  /// A context within the [MessageListPage] this action sheet was
+  /// triggered from.
+  final BuildContext pageContext;
+
+  void _onReactionPressed({
+    required EmojiCandidate emoji,
+    required bool isSelfVoted,
+  }) {
+    // Dismiss the enclosing action sheet immediately,
+    // for swift UI feedback that the user's selection was received.
+    Navigator.pop(pageContext);
+
+    final zulipLocalizations = ZulipLocalizations.of(pageContext);
+    doAddOrRemoveReaction(
+      context: pageContext,
+      doRemoveReaction: isSelfVoted,
+      messageId: message.id,
+      emoji: emoji,
+      errorDialogTitle: isSelfVoted
+        ? zulipLocalizations.errorReactionRemovingFailedTitle
+        : zulipLocalizations.errorReactionAddingFailedTitle);
   }
 
-  @override void onPressed() async {
-    String? errorMessage;
-    try {
-      await addReaction(PerAccountStoreWidget.of(pageContext).connection,
-        messageId: message.id,
-        reactionType: ReactionType.unicodeEmoji,
-        emojiCode: '1f44d',
-        emojiName: '+1',
-      );
-    } catch (e) {
-      if (!pageContext.mounted) return;
+  void _onMorePressed() {
+    // Dismiss current action sheet before opening emoji picker sheet.
+    Navigator.of(pageContext).pop();
 
-      switch (e) {
-        case ZulipApiException():
-          errorMessage = e.message;
-          // TODO(#741) specific messages for common errors, like network errors
-          //   (support with reusable code)
-        default:
-      }
+    showEmojiPickerSheet(pageContext: pageContext, message: message);
+  }
 
-      showErrorDialog(context: pageContext,
-        title: 'Adding reaction failed', message: errorMessage);
+  @override
+  Widget build(BuildContext context) {
+    assert(EmojiStore.popularEmojiCandidates.every(
+      (emoji) => emoji.emojiType == ReactionType.unicodeEmoji));
+
+    final zulipLocalizations = ZulipLocalizations.of(context);
+    final store = PerAccountStoreWidget.of(pageContext);
+    final designVariables = DesignVariables.of(context);
+
+    bool hasSelfVote(EmojiCandidate emoji) {
+      return message.reactions?.aggregated.any((reactionWithVotes) {
+        return reactionWithVotes.reactionType == ReactionType.unicodeEmoji
+          && reactionWithVotes.emojiCode == emoji.emojiCode
+          && reactionWithVotes.userIds.contains(store.selfUserId);
+      }) ?? false;
     }
+
+    return Container(
+      decoration: BoxDecoration(color: designVariables.contextMenuCancelBg),
+      child: Row(children: [
+        Flexible(child: Row(
+          spacing: 1,
+          children: List.unmodifiable(EmojiStore.popularEmojiCandidates.mapIndexed((index, emoji) {
```

Comment:
nit: pull this callback out as a helper method:
```suggestion
       children: List.unmodifiable(EmojiStore.popularEmojiCandidates
            .mapIndexed(_buildButton)))),
```

That'll dedent it and I think also help simplify understanding the structure.